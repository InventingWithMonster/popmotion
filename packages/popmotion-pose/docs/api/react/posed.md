---
title: Posed component
description: Create a posed component
category: react
---

# `posed`

`posed` is used to create animated and interactive components that you can reuse throughout your React site.

## Install

```bash
npm install react-pose --save
```

## Import

```javascript
import posed from 'react-pose'
```

## Usage

### Create a posed component

`pose` can be used in two ways:
1. Via HTML & SVG element aliases (eg; `posed.div()`)
2. Called directly with a React component (eg; `posed(MyComponent)`)

#### HTML & SVG element aliases

`pose` isn't called directly, instead we pass [posed props](/pose/api/props) to `posed.div`, `posed.button` etc. Every HTML and SVG element is supported:

```javascript
const DraggableCircle = posed.circle({
  draggable: 'x',
  dragBounds: { left: 0, right: 100 }
})

export default ({ radius }) => <DraggableCircle r={radius} />
```

#### Calling `pose` with a React component

`pose` is called directly, passing in [posed props](/pose/api/props):

```javascript
class MyComponent extends React.Component {
  /* ... */
}

const DraggableComponent = posed(MyComponent)({
  draggable: 'x',
  dragBounds: { left: 0, right: 100 }
})

export default ({ radius }) => <DraggableComponent r={radius} />
```

**Note:** React Pose requires a DOM element to correctly calculate animations,
as such it needs a reference to a component's underlaying DOM element. A custom
component can provide this via the callback `hostRef`:

```javascript
class MyComponent extends React.Component {
  render() {
    <div ref={this.props.hostRef}>Drag me</div>
  }
}
```

**Note 2:** `hostRef` is aliased to `innerRef` for components which support it,
however be warned that Styled Components [do not pass through `innerRef` in some
cases](https://github.com/styled-components/styled-components/issues/618#issuecomment-289247967)

### Set a pose

Poses can be set via the `pose` property. This can either be a string, or an array of strings to reference multiple poses:

```javascript
const Sidebar = posed.nav({
  open: { x: '0%' },
  closed: { x: '-100%' }
})

export default ({ isOpen }) => <Sidebar pose={isOpen ? 'open' : 'closed'} />
```

### Children

Using a posed component creates a new tree of posed components. Any children that are also posed components are automatically added to this tree.

This means that orchestrating animations through React trees becomes trivial, as a pose only has to be set on a parent. Any children with that pose defined will also animate:

```javascript
const Sidebar = posed.nav({
  open: { x: '0%', staggerChildren: 100 },
  closed: { x: '-100%' }
})

const NavItem = posed.li({
  open: { opacity: 1 },
  closed: { opacity: 0 }
})

export default ({ isOpen, navItems }) => (
  <Sidebar pose={isOpen ? 'open' : 'closed'}>
    <ul>
      {navItems.map(({ url, name }) => (
        <NavItem>
          <a href={url}>{name}</a>
        </NavItem>
      ))}
    </ul>
  </Sidebar>
)
```

<CodePen id="MVQepE" />

In tandem with the [`PoseGroup`](/pose/api/posegroup) component, this capability can be used to orchestrate sophisticated `enter` and `exit` animations.

This behaviour can be overridden by passing `newTree` to a posed component, which will ignore any parent posed components and create a new tree.

### Styling

Posed components are normal components, so they can be used with any CSS styling solution. For instance:

#### Styled Components

```javascript
const sidebarProps = {
  open: { x: '0%' },
  closed: { x: '-100%' }
}

const Sidebar = posed(styled.nav`
  width: 300px;
  background: red;
`)(sidebarProps);
```

#### `className`

```javascript
() => <Sidebar pose="closed" className="my-class" />
```

## Props

### `pose?: string | string[]`

The name or names of the current pose.

### `withParent?: boolean = true`

If set to `false`, this component won't subscribe to its parent posed component and create root for any further child components.

### `onPoseComplete?: Function`

A callback that fires whenever a pose has finished transitioning.

### `onChange?: { [key: string]: any }`

**Deprecated:** See `onValueChange`

### `onValueChange?: { [key: string]: any }`

`onValueChange` is a map of functions, each corresponding to a value being animated by the posed component and will fire when that value changes.

### `...props: { [key: string]: any }`

Remaining props will be provided to a pose's `transition` function when that pose is entered.
