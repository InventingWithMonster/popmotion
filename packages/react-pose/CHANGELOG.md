# Changelog

React Pose adheres to [Semantic Versioning](http://semver.org/).

## [1.5.3] 2018-05-04

### Fixed

- TS errors with PoseElementProps

## [1.5.2] 2018-04-13

### Fixed

- Checking `ref` for `instanceof Element` before setting in case `innerRef` has been passed to a component.

## [1.5.1] 2018-04-12

### Added

- `onDragStart`/`onDragEnd` callbacks to posed components.

## [1.5.0] 2018-04-12

### Added

- `posed(Component)` via [Jess Telford](https://twitter.com/jesstelford)
- React < 16.3 compatibility

## [1.4.0] 2018-04-11

### Added

- Labels (Pose 1.3.0)

## [1.3.0] 2018-04-10

### Added

- `beforeChildren` and `afterChildren` props (Pose 1.2.0)

## [1.2.0] 2018-04-09

### Added

- Dynamic pose props (Pose 1.1.0)

## [1.1.4] 2018-04-08

### Fixed

- Fixing botched React version mismatch in global file.

## [1.1.3] 2018-04-08

### Fixed

- [#285](https://github.com/Popmotion/popmotion/issues/285) PoseGroup child `enter` poses now animate correctly.

## [1.1.2] 2018-04-03

### Deprecated

- Deprecating `onChange` in favour of `onValueChange`. `onChange` will be removed in version 2.0.0. This is because React elements already use the `onChange` prop.

## [1.1.1] 2018-04-02

### Changed

- Using React `16.3.0` final release.

## [1.1.0] 2018-04-02

### Added

- Support for conditional rendering within `PoseGroup` (@efimweb)

## [1.0.2] 2018-03-29

### Fix

- Fixing types for `posed`

## [1.0.1] 2018-03-28

### Update

- Latest version of Pose

## [1.0.0] 2018-03-28

### Update

- Converting `PoseElement` from deprecated lifecycle method to `UNSAFE_componentWillUpdate`
- Fixing `posed` typescript def

## [0.1.7] 2018-03-25

### Update

- Updating Pose

## [0.1.6] 2018-03-25

### Update

- Updating Pose

## [0.1.5] 2018-03-25

### Update

- Updating Pose

## [0.1.4] 2018-03-24

### Update

- Exporting `popmotion` when using script import.

## [0.1.2] 2018-03-21

### Update

- Pose updated to 1.0.10

## [0.1.1] 2018-03-21

### Update

- Pose updated to 1.0.9

## [0.1.0] 2018-03-21

### Update

- RC

## [0.0.1] 2018-03-07

### Update

- Initial release