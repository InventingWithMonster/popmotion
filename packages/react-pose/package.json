{
  "name": "react-pose",
  "version": "2.0.0",
  "description": "A declarative animation library for React",
  "main": "./lib/index.js",
  "module": "./es/index.js",
  "types": "./types/index.d.ts",
  "scripts": {
    "build": "rollup -c && npm run measure",
    "postbuild": "babel es --out-dir es --no-babelrc --plugins annotate-pure-calls",
    "watch": "rollup -c -w",
    "test": "echo \"Error: no test specified\" && exit 1",
    "measure": "gzip -c $npm_package_unpkg | wc -c",
    "prettier": "prettier ./src/*",
    "prepublishOnly": "npm run prettier && npm run build"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/popmotion/packages/react-pose"
  },
  "keywords": [
    "animation",
    "dom",
    "declarative",
    "popmotion",
    "react",
    "react animation"
  ],
  "author": "Matt Perry",
  "license": "MIT",
  "presets": [
    "env"
  ],
  "peerDependencies": {
    "react": "^16.3.2",
    "react-dom": "^16.3.2",
    "popmotion": "^8.2.1"
  },
  "devDependencies": {
    "babel-cli": "^6.26.0",
    "babel-core": "^6.26.3",
    "babel-plugin-annotate-pure-calls": "^0.2.2",
    "prettier": "1.11.1",
    "react": "^16.3.2",
    "react-dom": "^16.3.2",
    "rollup": "^0.59.1",
    "rollup-plugin-commonjs": "^9.1.3",
    "rollup-plugin-node-resolve": "^3.3.0",
    "rollup-plugin-replace": "^2.0.0",
    "rollup-plugin-typescript2": "^0.14.0",
    "rollup-plugin-uglify": "^3.0.0",
    "typescript": "^2.7.2"
  },
  "dependencies": {
    "hey-listen": "^1.0.5",
    "tslib": "^1.9.1",
    "popmotion-pose": "^2.0.0"
  },
  "unpkg": "./dist/react-pose.js",
  "prettier": {
    "parser": "typescript",
    "singleQuote": true
  }
}
