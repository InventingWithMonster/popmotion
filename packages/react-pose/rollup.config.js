import typescript from 'rollup-plugin-typescript2';
import uglify from 'rollup-plugin-uglify';
import resolve from 'rollup-plugin-node-resolve';
import replace from 'rollup-plugin-replace';
import pkg from './package.json';

const typescriptConfig = { cacheRoot: 'tmp/.rpt2_cache' };
const declarationsConfig = Object.assign({}, typescriptConfig, {
  tsconfigOverride: { compilerOptions: { declaration: true, declarationDir: 'types' } }
});

const makeExternalPredicate = externalArr => {
  if (externalArr.length === 0) {
    return () => false;
  }
  const pattern = new RegExp(`^(${externalArr.join("|")})($|/)`);
  return id => pattern.test(id);
};

const deps = Object.keys(pkg.dependencies || {})
const peerDeps = Object.keys(pkg.peerDependencies || {})

const config = {
  input: { index:'src/index.ts', base: 'src/base.ts' },
  external: makeExternalPredicate(deps.concat(peerDeps))
};

const umd = Object.assign({}, config, {
  input: 'src/global.ts',
  output: {
    file: 'dist/react-pose.dev.js',
    format: 'umd',
    name: 'pose',
    exports: 'named',
    globals: { react: 'React' }
  },
  external: makeExternalPredicate(peerDeps),
  plugins: [
    typescript(typescriptConfig),
    resolve(),
    replace({
      'process.env.NODE_ENV': JSON.stringify('development')
    })
  ]
});

const umdProd = Object.assign({}, umd, {
  output: Object.assign({}, umd.output, {
    file: 'dist/react-pose.js'
  }),
  plugins: [
    typescript(typescriptConfig),
    resolve(),
    replace({
      'process.env.NODE_ENV': JSON.stringify('production')
    }),
    uglify(),
  ]
});

const es = Object.assign({}, config, {
  experimentalCodeSplitting: true,
  output: {
    dir: 'es',
    format: 'es',
    exports: 'named',
    entryNames: '[alias].js',
  },
  plugins: [typescript(typescriptConfig)]
});

const cjs = Object.assign({}, config, {
  experimentalCodeSplitting: true,
  output: {
    dir: 'lib',
    format: 'cjs',
    exports: 'named',
    entryNames: '[alias].js'
  },
  // plugins: [typescript(declarationsConfig)]
  plugins: [typescript(typescriptConfig)]
});

export default [/*umd, umdProd,*/ es, cjs];
