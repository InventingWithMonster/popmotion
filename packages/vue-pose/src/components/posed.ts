import Vue from 'vue';
import pose from 'popmotion-pose';
import supportedElements from './utils/supported-elements';
import {
  PosedComponentFactoryMap,
  PosedComponentFactoryFactory,
  PropsDefinition
} from './types';

const props: PropsDefinition = {
  pose: {
    type: String,
    default: 'init'
  }
};

const watch = {
  pose(newPose: string, prevPose: string) {
    if (newPose !== prevPose) {
      this.poser.set(newPose);
    }
  }
};

function destroyed() {
  this.poser.destroy();
}

const createPosedComponentFactory: PosedComponentFactoryFactory = el => (
  config = {}
) =>
  Vue.extend({
    props,
    mounted() {
      this.poser = pose(this.$el, {
        ...config,
        initialPose: this.$props.pose
      });
    },
    watch,
    destroyed,
    template: `<${el}><slot /><${el}/>`
  });

const posed: PosedComponentFactoryMap = supportedElements.reduce(
  (acc: PosedComponentFactoryMap, el) => {
    acc[el] = createPosedComponentFactory(el);

    return acc;
  },
  {}
);

export default posed;
