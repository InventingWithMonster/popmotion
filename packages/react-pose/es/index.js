import posed from './base.js';
export { PoseGroup } from './base.js';
import 'tslib';
import 'react';
import 'popmotion-pose';
import 'hey-listen';

var supportedElements = ['a', 'article', 'aside', 'audio', 'b', 'blockquote', 'body', 'br', 'button', 'canvas', 'caption', 'cite', 'code', 'col', 'colgroup', 'data', 'datalist', 'dialog', 'div', 'em', 'embed', 'fieldset', 'figcaption', 'figure', 'footer', 'form', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'head', 'header', 'hgroup', 'hr', 'i', 'iframe', 'img', 'input', 'label', 'legend', 'li', 'nav', 'object', 'ol', 'option', 'p', 'param', 'picture', 'pre', 'progress', 'q', 'section', 'select', 'span', 'strong', 'table', 'tbody', 'td', 'textarea', 'tfoot', 'th', 'thead', 'time', 'title', 'tr', 'ul', 'video', 'circle', 'clipPath', 'defs', 'ellipse', 'g', 'image', 'line', 'linearGradient', 'mask', 'path', 'pattern', 'polygon', 'polyline', 'radialGradient', 'rect', 'stop', 'svg', 'text', 'tspan'];

var index = /*#__PURE__*/supportedElements.reduce(function (acc, key) {
    acc[key] = posed(key);
    return acc;
}, posed);

export default index;