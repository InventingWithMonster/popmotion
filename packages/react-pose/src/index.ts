import posed from './base';
import supportedElements from './utils/supported-elements';

export * from './base';

export default supportedElements.reduce((acc, key) => {
  acc[key] = posed(key);
  return acc;
}, posed);


