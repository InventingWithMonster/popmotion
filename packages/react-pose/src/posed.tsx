import * as React from 'react';
import { ReactElement } from 'react';
import { PoseElement, PoseParentContext } from './components/PoseElement';
import {
  PoseElementProps,
  PoseContextProps
} from './components/PoseElement.types';
import { DomPopmotionConfig } from 'popmotion-pose/lib/types';

// TODO: Something is wrong with the TypeScript React bindings here,
// I've had to typecast a component as a Component a couple times now.
const PoseElementComponent = PoseElement as React.PureComponent;

export type ComponentFactory = (
  poseConfig?: DomPopmotionConfig
) => (props: PoseElementProps) => ReactElement<any>;

export type Posed = {
  (component: React.Component): ComponentFactory;
  [key: string]: ComponentFactory;
};

const componentCache = new Map<string | React.Component, ComponentFactory>();

const createComponentFactory = (key: string | React.Component) => {
  const componentFactory: ComponentFactory = (poseConfig = {}) => ({
    withParent = true,
    ...props
  }) =>
    !withParent || props.parentValues ? (
      <PoseElementComponent
        poseConfig={poseConfig}
        elementType={key}
        {...props}
      />
    ) : (
      <PoseParentContext.Consumer>
        {(parentCtx: PoseContextProps) => (
          <PoseElementComponent
            poseConfig={poseConfig}
            elementType={key}
            {...props}
            {...parentCtx}
          />
        )}
      </PoseParentContext.Consumer>
    );

  componentCache.set(key, componentFactory);

  return componentFactory;
};

const getComponentFactory = (key: string | React.Component) =>
  componentCache.has(key)
    ? componentCache.get(key)
    : createComponentFactory(key);

const posed = ((component: React.Component | string) =>
  getComponentFactory(component)) as Posed;

export default posed;
