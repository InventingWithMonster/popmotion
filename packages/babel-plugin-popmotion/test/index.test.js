import { transform as babel6Transform } from 'babel-core'
import { transform as babel7Transform } from '@babel/core'
import makeCases from 'jest-in-case'
import plugin from '../src'

const cases = {
  'rewritten import': {
    code: `
      import posed from 'react-pose'
    `
  },
  'default posed': {
    code: `
      import posed from 'react-pose'

      const AnimatedDiv = posed.div({ foo: 'bar' })
    `
  },
  'default custom name': {
    code: `
      import myPosed from 'react-pose'

      const AnimatedDiv = myPosed.div({ foo: 'bar' })
    `
  },
  'default as named specifier': {
    code: `
      import { PoseGroup, default as myPosed } from 'react-pose'

      const AnimatedDiv = myPosed.div({ foo: 'bar' })
    `
  },
  'different member expression': {
    code: `
      import test from 'test'

      const AnimatedDiv = test.div({ foo: 'bar' })
    `
  },
  'non member expression': {
    code: `
      import posed from 'react-pose'

      const AnimatedDiv = posed('div')({ foo: 'bar' })
      someCall(posed)
      const obj = { posed }
    `
  },
}

const createTester = transform => ({ code }) => {
  expect(
    transform(code, {
      plugins: [plugin],
      babelrc: false
    }).code
  ).toMatchSnapshot()
}

makeCases('babel 6', createTester(babel6Transform), cases)
makeCases('babel 7', createTester(babel7Transform), cases)
