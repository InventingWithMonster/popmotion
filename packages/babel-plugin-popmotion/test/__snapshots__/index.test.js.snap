// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel 6 default as named specifier 1`] = `
"
import { PoseGroup, default as myPosed } from 'react-pose/base';

const AnimatedDiv = /*#__PURE__*/myPosed('div')({ foo: 'bar' });"
`;

exports[`babel 6 default custom name 1`] = `
"
import myPosed from 'react-pose/base';

const AnimatedDiv = /*#__PURE__*/myPosed('div')({ foo: 'bar' });"
`;

exports[`babel 6 default posed 1`] = `
"
import posed from 'react-pose/base';

const AnimatedDiv = /*#__PURE__*/posed('div')({ foo: 'bar' });"
`;

exports[`babel 6 different member expression 1`] = `
"
import test from 'test';

const AnimatedDiv = test.div({ foo: 'bar' });"
`;

exports[`babel 6 non member expression 1`] = `
"
import posed from 'react-pose/base';

const AnimatedDiv = posed('div')({ foo: 'bar' });
someCall(posed);
const obj = { posed };"
`;

exports[`babel 6 rewritten import 1`] = `
"
import posed from 'react-pose/base';"
`;

exports[`babel 7 default as named specifier 1`] = `
"import { PoseGroup, default as myPosed } from \\"react-pose/base\\";
const AnimatedDiv =
/*#__PURE__*/
myPosed(\\"div\\")({
  foo: 'bar'
});"
`;

exports[`babel 7 default custom name 1`] = `
"import myPosed from \\"react-pose/base\\";
const AnimatedDiv =
/*#__PURE__*/
myPosed(\\"div\\")({
  foo: 'bar'
});"
`;

exports[`babel 7 default posed 1`] = `
"import posed from \\"react-pose/base\\";
const AnimatedDiv =
/*#__PURE__*/
posed(\\"div\\")({
  foo: 'bar'
});"
`;

exports[`babel 7 different member expression 1`] = `
"import test from 'test';
const AnimatedDiv = test.div({
  foo: 'bar'
});"
`;

exports[`babel 7 non member expression 1`] = `
"import posed from \\"react-pose/base\\";
const AnimatedDiv = posed('div')({
  foo: 'bar'
});
someCall(posed);
const obj = {
  posed
};"
`;

exports[`babel 7 rewritten import 1`] = `"import posed from \\"react-pose/base\\";"`;
